File: ios-app-wip/MobileClaw/Memory/MemoryRepository.swift
import Foundation

// 1:1 port of android-app/.../memory/MemoryRepository.kt — the protocol the
// orchestration module talks to. Two implementations: InMemory (eval +
// tests, no persistence) and SwiftData-backed (production).

public protocol MemoryRepository: Sendable {
    func save(episode: Episode) async
    func save(repair: RepairRecord) async
    func saveDeviceFact(key: String, value: String, sourceEpisodeId: String?) async
    func recallForPlanning(userMessage: String, tokenBudget: Int) async -> String
    func recallRepairs(skillName: String, error: String, limit: Int) async -> [RepairRecord]
    func getDeviceFacts() async -> [DeviceFact]
    func evictIfNeeded(maxSizeBytes: Int64) async
    func clearAll() async
}

public extension MemoryRepository {
    func recallForPlanning(userMessage: String) async -> String {
        await recallForPlanning(userMessage: userMessage, tokenBudget: 2048)
    }
    func recallRepairs(skillName: String, error: String) async -> [RepairRecord] {
        await recallRepairs(skillName: skillName, error: error, limit: 3)
    }
    func evictIfNeeded() async {
        await evictIfNeeded(maxSizeBytes: 100 * 1024 * 1024)
    }
}

File: android-app/app/src/main/java/com/edgecat/app/memory/MemoryRepository.kt
package com.edgecat.app.memory

interface MemoryRepository {
  suspend fun saveEpisode(episode: Episode)
  suspend fun saveRepair(repair: RepairRecord)
  suspend fun saveDeviceFact(key: String, value: String, sourceEpisodeId: String? = null)
  suspend fun recallForPlanning(userMessage: String, tokenBudget: Int = 2048): String
  suspend fun recallRepairs(skillName: String, error: String, limit: Int = 3): List<RepairRecord>
  suspend fun getDeviceFacts(): List<DeviceFact>
  suspend fun evictIfNeeded(maxSizeBytes: Long = 100L * 1024 * 1024)
  suspend fun clearAll()
}
